express-notes-api/index.js

```diff
@@ -0,0 +1,95 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+const dataJSON = require('./data.json');
+
+app.get('/api/notes', (req, res) => {
+  const notesJSON = [];
+  for (const key in dataJSON.notes) {
+    notesJSON.push(dataJSON.notes[key]);
+  }
+  if (notesJSON.length > 0) {
+    res.status(200).send(notesJSON);
+  } else {
+    res.send(notesJSON);
+  }
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const id = Number(req.params.id);
+  if (id < 0 || isNaN(id)) {
+    const message400 = { error: 'id must be a positive integer' };
+    res.status(400).send(message400);
+  } else if (!(dataJSON.notes[id])) {
+    const message404 = { error: `cannot find note with id ${id}` };
+    res.status(404).send(message404);
+  } else {
+    res.status(200).send(dataJSON.notes[id]);
+  }
+});
+
+app.use(express.json());
+
+app.post('/api/notes', (req, res) => {
+  if (!(req.body.content) || req.body.content === null) {
+    const message400 = { error: 'content is a required field.' };
+    res.status(400).send(message400);
+  } else if (req.body.content !== null) {
+    const newContent = req.body;
+    const id = dataJSON.nextId++;
+    newContent.id = id;
+    dataJSON.notes[id] = newContent;
+    res.status(201).send(dataJSON.notes[id]);
+    const newObj = JSON.stringify(dataJSON, null, 2);
+    fs.writeFile('./data.json', newObj, err => {
+      const message500 = { error: 'An unexpected error occured.' };
+      if (err) console.error(res.status(500).send(message500));
```

Comment:
We want to split this into two separate operations
1. log the error object
2. send the 500 response

```js
console.error(err);
res.status(500).send(message500);
```